scuba/config.py

```diff
@@ -323,21 +323,49 @@ def _get_delimited_str_list(data: CfgData, key: str, sep: str) -> List[str]:
     return s.split(sep) if s else []
 
 
-def _get_volumes(data: CfgData) -> Optional[Dict[Path, ScubaVolume]]:
+def _get_volumes(
+    data: CfgData, scuba_root: Optional[Path]
+) -> Optional[Dict[Path, ScubaVolume]]:
     voldata = _get_dict(data, "volumes")
     if voldata is None:
         return None
 
     vols = {}
     for cpath_str, v in voldata.items():
-        cpath = _expand_path(cpath_str)
-        vols[cpath] = ScubaVolume.from_dict(cpath, v)
+        cpath = _expand_path(cpath_str)  # no base_dir; container path must be absolute.
+        vols[cpath] = ScubaVolume.from_dict(cpath, v, scuba_root)
     return vols
 
 
-def _expand_path(in_str: str) -> Path:
+def _expand_path(in_str: str, base_dir: Optional[Path] = None) -> Path:
+    """Expand variables in a path string and make it absolute.
+
+    Environment variable references (e.g. $FOO and ${FOO}) are expanded using
+    the host environment.
+
+    After environment variable expansion, absolute paths are returned as-is.
+    Relative paths must start with ./ or ../ and are joined to base_dir, if
+    provided.
```

Comment:
I made up a stupid test to ensure that variable expansion happens before the relative path checking, and it passed! Up to you whether you want to add it to the test suite or not to prevent regressions:
```diff
diff --git a/tests/test_config.py b/tests/test_config.py
index 3b7fd43..ab2b4d6 100644
--- a/tests/test_config.py
+++ b/tests/test_config.py
@@ -1104,6 +1104,27 @@ class TestConfig:
             )
         self._invalid_config("Relative path must start with ./ or ../")
 
+    def test_volumes_hostpath_rel_in_env(
+        self, monkeypatch, in_tmp_path
+    ) -> None:
+        """volume definitions can contain environment variables, including relative path portions"""
+        monkeypatch.setenv("PREFIX", "./")
+        with open(".scuba.yml", "w") as f:
+            f.write(
+                r"""
+                image: na
+                volumes:
+                  /foo: ${PREFIX}/foo
+                """
+            )
+
+        config = load_config()
+        vols = config.volumes
+        assert vols is not None
+        assert len(vols) == 1
+
+        assert_vol(vols, "/foo", in_tmp_path / "foo")
+
     def test_volumes_contpath_rel(self, monkeypatch, in_tmp_path) -> None:
         with open(".scuba.yml", "w") as f:
             f.write(
```

Reply:
Thanks!

```console
$ git apply will.diff
$ ./code_format.py --fix
$ git commit --author='Will Sloan <will@848.email>' --date='2023-09-15T03:13:05Z
[volume-relative-path d4a61bc] tests: Add test ensuring variable expansion happens before relative path check                                                                                                                                
 Author: Will Sloan <will@848.email>                                                                                                                                                                                                         
 Date: Fri Sep 15 03:13:05 2023 +0000                                                                                                                                                                                                        
 1 file changed, 21 insertions(+)
```

👉 d4a61bcb43d008d14813db605ed754f5143f378b